gladier/flow.py

```diff
@@ -0,0 +1,261 @@
+        self,
+        running_flows: t.List[GladierFlowRun],
+        run_completion_callback: t.Optional[t.Callable[[GladierFlowRun], bool]] = None,
+    ) -> bool:
+        complete_runs: t.List[GladierFlowRun] = []
+        for running_flow in running_flows:
+            if running_flow.is_complete():
+                complete_runs.append(running_flow)
+                if run_completion_callback is not None:
+                    end_runs = run_completion_callback(running_flow)
+                    if end_runs:
+                        return end_runs
+        for complete_run in complete_runs:
+            running_flows.remove(complete_run)
+        return False
+
+    def run_many(
+        self,
+        run_input_iterator: t.Iterable[JSONObject],
+        max_concurrency=5,
+        min_delay=100,
+        run_completion_callback: t.Optional[t.Callable[[GladierFlowRun], bool]] = None,
+        tags: t.Optional[t.List[str]] = None,
+        label_template: t.Optional[str] = None,
+        label_template_values: t.Optional[t.List[str]] = None,
+        wait_for_all_completions=False,
+    ):
+        running_flows: t.List[GladierFlowRun] = []
+        last_run_time = time.time()
+        end_runs = False
+        if label_template_values is None:
+            label_template_values = []
+
+        for run_input in run_input_iterator:
+            at_concurrency_limit = True
+            while at_concurrency_limit:
+                end_runs = self.update_running_flows(
+                    running_flows, run_completion_callback=run_completion_callback
+                )
+                at_concurrency_limit = len(running_flows) >= max_concurrency
+                if at_concurrency_limit and not end_runs:
+                    time.sleep(2)
+
+            if end_runs:
+                break
+
+            if label_template is not None:
+                evaluated_template_vals: t.List[t.Any] = []
+                for label_template_value in label_template_values:
+                    evaluated_template_vals.append(
+                        eval_jsonpath_for_input(label_template_value, run_input)
+                    )
+                run_label = label_template.format(*evaluated_template_vals)
+            else:
+                run_label = f"Run of {self.flow_id}"
+
+            now = time.time()
+            sleep_time = (last_run_time + min_delay) - now
+            if sleep_time > 0:
+                time.sleep(float(sleep_time) / 1000.0)
+            flow_run = self.run_flow(run_input, run_label, tags)
+            last_run_time = time.time()
+            running_flows.append(flow_run)
+
+        while wait_for_all_completions and len(running_flows) > 0:
+            wait_for_all_completions = not self.update_running_flows(
+                running_flows, run_completion_callback=run_completion_callback
+            )
+            if wait_for_all_completions and len(running_flows) > 0:
+                time.sleep(2)
```

Comment:
The `flow.py` file extends Gladier functionality, but doesn't use any of the internals and instead uses the external flows client for storing ids. This needs to be updated to use the internals instead. It may be easier to remove and add these features at a later time.